### PR TITLE
Add Cppcheck Static Code Analysis to CI Workflow

### PR DESCRIPTION
**Pull request by ThibFrgsGmz**
| | |
|:---|:---|
|**_Originating Project/Creator_**| |
|**_Affected Component_**| CI |
|**_Affected Architectures(s)_**| None |
|**_Related Issue(s)_**| None |
|**_Has Unit Tests (y/n)_**| N/A |
|**_Builds Without Errors (y/n)_**| N/A |
|**_Unit Tests Pass (y/n)_**| N/A |
|**_Documentation Included (y/n)_**| n |

---
## Change Description

This PR aims to add a new workflow "Cppcheck Scan" that runs the cppcheck linter on the framework c++ code base.
The results are displayed in the workflow summary in the Actions tab.

## Rationale

Add more static analysis to catch hidden bugs that other analyzers would not find.

## Testing/Review Recommendations

This workflow is largely adapted from the NASA cFS framework [static analysis workflow](https://github.com/nasa/cFS/blob/c36aa2c1df0fb47a3838577908af3d0d0ab0ef54/.github/workflows/static-analysis.yml).

I'll let you change the triggering policy of this workflow (on push, pull request etc.).

## Future Work

I would like to add scanning for MISRA rules, perhaps as a new step in this workflow.

### CODE REVIEW

**Comment by LeStarch** [comment]
We need to fix the spelling errors.  I'll let the rest run.

File: .github/scripts/cppcheck-xml2text.xslt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+| error | warning | style | performance | portability | information |
+| --- | --- | --- | --- | --- | --- |
+| <xsl:value-of select="count(//error[@severity='error'])"/> | <xsl:value-of select="count(//error[@severity='warning'])"/> | <xsl:value-of select="count(//error[@severity='style'])"/> | <xsl:value-of select="count(//error[@severity='performance'])"/> | <xsl:value-of select="count(//error[@severity='portability'])"/> | <xsl:value-of select="count(//error[@severity='information'])"/> |
+
+| severity | location | error id | issue |
+| --- | --- | --- | --- |
+<xsl:for-each select="results//error">| <xsl:value-of select="@severity"/> | <xsl:value-of select="location/@file"/>:<xsl:value-of select="location/@line"/> | <xsl:value-of select="@id"/> | <xsl:value-of select="@msg"/> |
+</xsl:for-each>
+</xsl:if>
+**<xsl:value-of select="count(//error)"/> error(s) reported**
```
Comment (thomas-bc) [inline .github/scripts/cppcheck-xml2text.xslt:15]: See my comment below. Maybe we could change this to
```
**<xsl:value-of select="count(//error)"/> alerts(s) reported**
**<xsl:value-of select="count(//error[@severity='error'])"/> error(s) reported**
```
That way we count the number of actual `severity=error` in the last step of the workflow.

**Comment by LeStarch** [comment]
I looked through the results, and it has made some good catches!  However, most of this is false positives, or at least I would argue they are.  Can we filter/remove some results?

Most of the flagged items are on local storage buffers not being initialized in constructor....but we don't care as the memory isn't used until layer.  It is extra performance cost to zero the arrays in the constructor.

**Review by thomas-bc — CHANGES_REQUESTED** [review]
This looks good! I requested a few changes, notably to make the check fail only if it finds `Error`s, and not on all type of alerts.

File: .github/workflows/cppcheck-scan.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# Adapted from https://github.com/nasa/cFS/blob/c36aa2c1df0fb47a3838577908af3d0d0ab0ef54/.github/workflows/static-analysis.yml
+name: "Cppcheck Scan"
+
+on:
+  push:
+    branches:
+      - "**"
```
[the remark this answers is not in the document]
Comment (ThibFrgsGmz) [reply .github/workflows/cppcheck-scan.yml]: Done!

File: .github/scripts/cppcheck-xml2text.xslt (hunk 1 above)
Comment (ThibFrgsGmz) [reply .github/scripts/cppcheck-xml2text.xslt]: Awesome, thanks for the suggestion!

File: .github/workflows/cppcheck-scan.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          checkout_path: ${{ github.workspace }}/source
+          category: "cppcheck"
+
+      - name: Archive static analysis artifacts to download and view
+        uses: actions/upload-artifact@v3
+        with:
+          name: cppcheck-errors
+          path: ./*cppcheck_err.*
+
+      - name: Check for reported errors
```
[the remark this answers is not in the document]
Comment (ThibFrgsGmz) [reply .github/workflows/cppcheck-scan.yml:67]: Comment added w/ link to SO!

File: .github/workflows/cppcheck-scan.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      # referencing the compile_commands.json file produced by the tool.  This will capture the correct include paths and
+      # compile definitions based on how the source is actually compiled. See https://cppcheck.sourceforge.io/manual.html
+      - name: Generate & build F´
+        run: |
+          fprime-util generate -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
+          fprime-util build --all --jobs "$(nproc || printf '%s\n' 1)"
+          echo CPPCHECK_OPTS=--project="$GITHUB_WORKSPACE/build-fprime-automatic-native/compile_commands.json" >> $GITHUB_ENV
+
+      - name: Run cppcheck
+        run: cppcheck --force --inline-suppr --std=c++11 -j "$(nproc || printf '%s\n' 1)" --max-ctu-depth=16 --enable=warning,performance,portability --suppress=variableScope --inconclusive --xml $CPPCHECK_OPTS 2> cppcheck_err.xml
```
[the remark this answers is not in the document]
Comment (ThibFrgsGmz) [reply .github/workflows/cppcheck-scan.yml]: I added it, it seems to have worked, right?

[the remark this answers is not in the document]
Comment (thomas-bc) [reply .github/workflows/cppcheck-scan.yml]: It worked! I see the alerts inline in the Code Scanning tab. Thank you!

**Review by thomas-bc — COMMENTED** [review]
Looking good! I don't have any other comments. There's 2 spelling issues I mentioned though.

Have you been able to look into if/how CppCheck handles filters on results? Say if we want to filter out the "buffers not being initialized in constructor" alerts that @LeStarch mentioned above.

**Comment by ThibFrgsGmz** [comment]
@thomas-bc I had a quick look and it doesn't seem to be possible. About the `uninitMemberVar` warning, the creator of the tool doesn't seem to mention any filtering mechanism: https://stackoverflow.com/a/27473629/14873360.

Personally I think it's unsafe to try to hide/filter this kind of error, it's a warning that I think is legitimate to bring up. Admittedly, it's a false positive for the moment but we are not safe from a later error.

**Comment by thomas-bc** [comment]
There is still issues with the spellchecking for some odd reason. Could you try to have only `cppcheck` lowercase listed in the `expect.txt`, and remove `Cppcheck` and `CPPCHECK`? It looks like that might be the issue -- I think `expect.txt` is somewhat case insensitive. You can test in a different branch of yours to make sure that works before pushing here.

**Comment by ThibFrgsGmz** [comment]
@thomas-bc Done! Thanks for your help ! 👌

**Review by thomas-bc — APPROVED** [review]
Approved, thank you!

**Review by LeStarch — APPROVED** [review]
LGTM!

**Pull request merged**